src/navigation/post_backup.py

```diff
@@ -0,0 +1,113 @@
+from __future__ import annotations
+from dataclasses import dataclass
+import numpy as np
+from typing import Optional
+from state import BaseState
+from trajectory import Trajectory
+from aenum import Enum, NoAlias
+from context import Context
+from util.ros_utils import get_rosparam
+from util.np_utils import perpendicular_2d
+from shapely.geometry import Point, LineString
+from util.SE3 import SE3
+
+POST_RADIUS = get_rosparam("gate/post_radius", 0.7) * get_rosparam(
+    "single_fiducial/post_avoidance_multiplier", 4.0
+)  # add a big buffer to the post radius, 4 is somewhat arbitrary but it doesn't really matter
+BACKUP_DISTANCE = get_rosparam("recovery/recovery_distance", 1.0)
+STOP_THRESH = 0.2
+DRIVE_FWD_THRESH = 0.95
+
+
+@dataclass
+class AvoidPostTrajectory(Trajectory):
+    def avoid_post_trajectory(rover_pose: SE3, post_pos: np.ndarray, waypoint_pos: np.ndarray) -> AvoidPostTrajectory:
+        """
+        Generates a trajectory that avoids a post until the rover has a clear path to the waypoint
+        """
+        rover_pos = rover_pose.position
+        rover_direction = rover_pose.rotation.direction_vector()
+
+        # Converting to 2d arrays
+        post_pos = post_pos[:2]
+        rover_pos = rover_pos[:2]
+        rover_direction = rover_direction[:2]
+        waypoint_pos = waypoint_pos[:2]
+
+        # normalize rover_direction
+        rover_direction = rover_direction / np.linalg.norm(rover_direction)
+
+        # create a Shapeley point of raidus POST_RADIUS around the post
+        post_circle = Point(post_pos[0], post_pos[1]).buffer(POST_RADIUS)
+
+        # generate a point BACKUP_DISTANCE behind the rover
+        backup_point = rover_pos - BACKUP_DISTANCE * rover_direction
+
+        # generate a line from the backup point to the waypoint
+        path = LineString([backup_point, waypoint_pos])
+        coords: np.ndarray = np.array([])
+        # check if the path intersects the post circle
+        if path.intersects(post_circle):
+            # get a vector perpendicular to rover direction
+            left_perp = perpendicular_2d(rover_direction)  # (-y,x)
```

Comment:
I'm also a little confused about what this avoidance_point should look like and what vector to use. What would be the scenario where it doesn't avoid the post?

Reply:
If the rover is already facing perpendicular to the direction of the next waypoint, this could select on avoidance point that goes directly through the post I think. Could be wrong I haven't thought this through very much

Reply:
Should be resolved now. Its checking first to see if just backup point to waypoint is clear and only then it generates a new avoidance point as per the way the code is written